demo/client/backend/mock/server.go

```diff
@@ -86,23 +107,31 @@ func invoke(c *gin.Context) {
 	}
 
 	res := stub.MockInvoke("someTxID", args)
-	if res.Status != shim.OK {
-		//fmt.Printf("Chaincode error: %s", res.Message)
-		logger.Error(fmt.Sprintf("Chaincode Error: %s\n", res.Message))
-		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"message": res.Message})
-		return
+	logger.Debugf("invocation response: status='%v' / payload='%v' / message='%s'", res.Status, res.Payload, res.Message)
+
+	// we might get payload and response regardless of invocation success, so try to decode in all cases
+	var fpcResponse string
+	if res.Payload != nil {
+		var response utils.Response
+		// unwarp ecc response and return only responseData
+		// a proper client would now also verify response signature
+		if err = json.Unmarshal(res.Payload, &response); err == nil {
+			logger.Debugf("decoded fpc response: ResponseData='%s'",
+				response.ResponseData)
+			fpcResponse = string(response.ResponseData)
+		}
 	}
-
-	// unwarp ecc response and return only responseData
-	// a proper client would now also verify response signature
-	var response utils.Response
-	err = json.Unmarshal(res.Payload, &response)
+	// NOTE: we do return error even if the invocation get success back
+	// but does not contain a response payload. According to the auction
+	// specifications, all queries and transactions should return a response
+	// object (even more specifically, an object which at the very least
+	// contains a 'status' field)
 	if err != nil {
-		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
-		return
+		fpcResponse = fmt.Sprintf("{ \"status\": { \"rc\" : %d, \"message\" : \"No valid response payload received due to errror=%v\" } }",
```

Comment:
ok for returning the error, but forcing this status object does not belong here, it belongs to the chaincode.
If the contract is "there must be a status object" and eventually there isn't one, then the chaincode has a bug.
In addition, whoever ultimately parses the object, also has to check that there is such object, or raise an error otherwise. So it seems redundant to me to handle it here.

Reply:
Somewhat disagree: This could be a problem of the code between client and chaincode, not necessarily chaincode itself.  And as discussed so far the UI expects a status object in all cases, that was at least my understanding (also from discussion on slack).  Ultimately, i have no strong feeling whether non-chaincode errors should be a status object or some HTTP error, we just have to agree what we want and do it consistently for both mock and fabric-gateway client backends.  Note this is demo specific and right now consistent.  So i would propose to table the discussion on whether mabye mapping non-chaincode errrors to HTTP errors (and if so, which ones) would be better until the point where we try to generalize the mock backend and fabric-gateway code?